File: javascript/src/allowance.ts
/**
 * Configurable free-allowance reset window resolver (WS9a).
 *
 * Pure, UTC-only, DATE-granularity (time-of-day is discarded) window resolver.
 * MUST exactly match the Python resolver: this is the cross-language parity
 * baseline for `allowancePeriod` handling. All computation uses UTC methods
 * exclusively (`getUTCFullYear`, `Date.UTC`, ...) — never local-timezone
 * Date methods.
 *
 * `resolveCalendarWindow` is a separate, anchor-free resolver used by
 * per-feature invocation-count limits (`PlanDefinition.featureLimits`). It
 * supports `"daily"`/`"weekly"`/`"monthly"`/`"yearly"` cadences, all
 * calendar-aligned (every user resets at the same instant) rather than
 * anchored to a per-user timestamp — MUST exactly match
 * `bursar.allowance.resolve_calendar_window` in the Python SDK.
 */

/** Supported free-allowance reset window modes. */
export type AllowancePeriod = "calendar_month" | "rolling_30d" | "anniversary";

/** Supported per-feature invocation-count limit cadences. */
export type FeatureLimitPeriod = "daily" | "weekly" | "monthly" | "yearly";

const MS_PER_DAY = 86_400_000;

/** Truncate a `Date` to UTC midnight (DATE granularity, discards time-of-day). */
function utcMidnight(d: Date): Date {
  return new Date(Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()));
}

/** Number of days in the given UTC year/month (0-indexed month, like `Date.UTC`). */
function daysInUtcMonth(year: number, month: number): number {
  // Day 0 of "next month" is the last day of `month`.
  return new Date(Date.UTC(year, month + 1, 0)).getUTCDate();
}

/**
 * The `day`-th day of the UTC month `(year, month)`, clamped to that month's
 * actual length (e.g. day 31 in a 28-day February clamps to the 28th).
 */
function clampedDayInMonth(year: number, month: number, day: number): Date {
  const lastDay = daysInUtcMonth(year, month);
  return new Date(Date.UTC(year, month, Math.min(day, lastDay)));
}

/** First day of the UTC month containing `d`, at midnight UTC. */
function calendarMonthStart(d: Date): Date {
  return new Date(Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), 1));
}

function resolveCalendarMonth(now: Date): { start: Date; end: Date } {
  const start = calendarMonthStart(now);
  const end = new Date(Date.UTC(start.getUTCFullYear(), start.getUTCMonth() + 1, 1));
  return { start, end };
}

function resolveRolling30d(now: Date, anchor: Date | null): { start: Date; end: Date } {
  if (anchor === null) return resolveCalendarMonth(now);
  const anchorMidnight = utcMidnight(anchor);
  const nowMidnight = utcMidnight(now);
  const elapsedDays = Math.floor((nowMidnight.getTime() - anchorMidnight.getTime()) / MS_PER_DAY);
  const windowIndex = Math.floor(elapsedDays / 30);
  const start = new Date(anchorMidnight.getTime() + windowIndex * 30 * MS_PER_DAY);
  const end = new Date(start.getTime() + 30 * MS_PER_DAY);
  return { start, end };
}

function resolveAnniversary(now: Date, anchor: Date | null): { start: Date; end: Date } {
  if (anchor === null) return resolveCalendarMonth(now);
  const anchorMidnight = utcMidnight(anchor);
  const nowMidnight = utcMidnight(now);
  const resetDay = anchorMidnight.getUTCDate();

  // Candidate reset date in the same UTC year/month as `now`, clamped.
  const candidate = clampedDayInMonth(
    nowMidnight.getUTCFullYear(),
    nowMidnight.getUTCMonth(),
    resetDay,
  );

  let start: Date;
  if (candidate.getTime() <= nowMidnight.getTime()) {
    start = candidate;
  } else {
    // The clamped reset date this month is still ahead of `now` — the current
    // window started last month.
    start = clampedDayInMonth(
      nowMidnight.getUTCFullYear(),
      nowMidnight.getUTCMonth() - 1,
      resetDay,
    );
  }

  const end = clampedDayInMonth(start.getUTCFullYear(), start.getUTCMonth() + 1, resetDay);
  return { start, end };
}

/**
 * Resolve the free-allowance reset window containing `now` for the given
 * `period` mode, anchored (where applicable) at `anchor` (typically the
 * plan-assignment timestamp).
 *
 * `start`/`end` are UTC-midnight `Date`s at DATE granularity; `end` is
 * EXCLUSIVE (the window is `[start, end)`).
 */
export function resolveAllowanceWindow(
  now: Date,
  period: AllowancePeriod,
  anchor: Date | null,
): { start: Date; end: Date } {
  switch (period) {
    case "calendar_month":
      return resolveCalendarMonth(now);
    case "rolling_30d":
      return resolveRolling30d(now, anchor);
    case "anniversary":
      return resolveAnniversary(now, anchor);
    default:
      throw new Error(`unrecognized allowance period: ${String(period)}`);
  }
}

/** `[now, now + 1 day)` at UTC midnight. */
function dayWindow(now: Date): { start: Date; end: Date } {
  const start = utcMidnight(now);
  return { start, end: new Date(start.getTime() + MS_PER_DAY) };
}

/** Monday-start ISO week window: `[Monday, next Monday)`. */
function isoWeekWindow(now: Date): { start: Date; end: Date } {
  const midnight = utcMidnight(now);
  // getUTCDay(): 0=Sunday..6=Saturday. Days since the most recent Monday.
  const dayOfWeek = midnight.getUTCDay();
  const daysSinceMonday = (dayOfWeek + 6) % 7;
  const start = new Date(midnight.getTime() - daysSinceMonday * MS_PER_DAY);
  return { start, end: new Date(start.getTime() + 7 * MS_PER_DAY) };
}

/** `[Jan 1, next Jan 1)` of `now`'s UTC year. */
function yearWindow(now: Date): { start: Date; end: Date } {
  const start = new Date(Date.UTC(now.getUTCFullYear(), 0, 1));
  const end = new Date(Date.UTC(now.getUTCFullYear() + 1, 0, 1));
  return { start, end };
}

/**
 * Resolve the calendar-aligned `[start, end)` window for a per-feature
 * invocation-count limit cadence (`FeatureLimit.period`).
 *
 * Unlike {@link resolveAllowanceWindow}, this resolver takes no anchor: every
 * user resets at the same UTC instant. `start`/`end` are UTC-midnight `Date`s
 * at DATE granularity; `end` is EXCLUSIVE (the window is `[start, end)`).
 *
 * MUST exactly match `bursar.allowance.resolve_calendar_window` in Python
 * (same Monday-start ISO week, same year boundary).
 */
export function resolveCalendarWindow(
  now: Date,
  period: FeatureLimitPeriod,
): { start: Date; end: Date } {
  switch (period) {
    case "daily":
      return dayWindow(now);
    case "weekly":
      return isoWeekWindow(now);
    case "monthly":
      return resolveCalendarMonth(now);
    case "yearly":
      return yearWindow(now);
    default:
      throw new Error(`unrecognized feature-limit period: ${String(period)}`);
  }
}

File: python/src/ducto/allowance.py
"""Pure resolver for configurable free-allowance reset windows (WS9).

``resolve_allowance_window`` computes the ``[period_start, period_end)`` window
that a user's plan allowance resets on, given the allowance period mode and an
anchor timestamp (when the user was assigned the plan). All computation is
UTC-only, date granularity (time-of-day is discarded) — there is no local-time
conversion anywhere, so the resolver is unaffected by DST in any zone.

Supported ``period`` values (mirrors ``PlanDefinition.allowance_period``):

- ``"calendar_month"`` (default): resets on the 1st of each UTC calendar month.
  This is the regression-safety baseline — it must match the pre-WS9 SQL
  ``date_trunc('month', now() AT TIME ZONE 'UTC')`` behavior exactly.
- ``"rolling_30d"``: resets every 30 days from the anchor (plan-assignment time).
  Falls back to ``calendar_month`` when no anchor is available.
- ``"anniversary"``: resets monthly on the anchor's day-of-month, clamped to the
  target month's actual length (e.g. an anchor of the 31st resets on the 28th/29th
  in February, then returns to the 31st in a 31-day month). Falls back to
  ``calendar_month`` when no anchor is available.
"""

from __future__ import annotations

import calendar
from datetime import date, datetime, timedelta

__all__ = ["resolve_allowance_window"]


def _month_start(d: date) -> date:
    return d.replace(day=1)


def _add_months(d: date, months: int) -> date:
    """Add ``months`` calendar months to ``d``, returning the 1st of that month."""
    total = (d.year * 12 + (d.month - 1)) + months
    year, month0 = divmod(total, 12)
    return date(year, month0 + 1, 1)


def _calendar_month_window(now: date) -> tuple[date, date]:
    start = _month_start(now)
    end = _add_months(start, 1)
    return start, end


def _clamped_day(year: int, month: int, day: int) -> date:
    """Return ``date(year, month, day)``, clamping ``day`` to the month's length."""
    last_day = calendar.monthrange(year, month)[1]
    return date(year, month, min(day, last_day))


def _rolling_30d_window(now: date, anchor_date: date) -> tuple[date, date]:
    elapsed_days = (now - anchor_date).days
    window_index = elapsed_days // 30
    start = anchor_date + timedelta(days=30 * window_index)
    end = start + timedelta(days=30)
    return start, end


def _anniversary_window(now: date, anchor_date: date) -> tuple[date, date]:
    """Monthly window whose reset day-of-month equals ``anchor_date.day``.

    The reset day is clamped to each target month's actual length (e.g. an
    anchor day of 31 resets on the 28th/29th in February, then RETURNS to the
    31st in a 31-day month — clamping is re-evaluated per month, not "sticky").
    """
    anchor_day = anchor_date.day

    # Candidate reset date in `now`'s own month.
    this_month_reset = _clamped_day(now.year, now.month, anchor_day)

    if now >= this_month_reset:
        start = this_month_reset
        end_year, end_month = (now.year + 1, 1) if now.month == 12 else (now.year, now.month + 1)
        end = _clamped_day(end_year, end_month, anchor_day)
    else:
        # Most recent occurrence was in the previous month.
        prev_year, prev_month = (now.year - 1, 12) if now.month == 1 else (now.year, now.month - 1)
        start = _clamped_day(prev_year, prev_month, anchor_day)
        end = this_month_reset

    return start, end


def resolve_allowance_window(now: datetime, period: str, anchor: datetime | None) -> tuple[date, date]:
    """Resolve the ``[period_start, period_end)`` allowance window (UTC, date-only).

    Args:
        now: Current instant. Only the UTC date part is used (time-of-day is
            discarded); callers should pass a UTC-aware (or UTC-naive)
            datetime — this function does not perform any timezone conversion.
        period: One of ``"calendar_month"``, ``"rolling_30d"``, ``"anniversary"``.
        anchor: When the user was assigned the plan (used by ``rolling_30d`` and
            ``anniversary``). ``None`` falls back to ``calendar_month`` behavior
            for those modes.

    Returns:
        ``(period_start, period_end)`` as ``date`` objects, both UTC, with
        ``period_end`` EXCLUSIVE (i.e. the window is ``[period_start, period_end)``).

    Raises:
        ValueError: If ``period`` is not a recognized allowance-period mode.
    """
    now_date = now.date()

    if period == "calendar_month":
        return _calendar_month_window(now_date)

    if period == "rolling_30d":
        if anchor is None:
            return _calendar_month_window(now_date)
        return _rolling_30d_window(now_date, anchor.date())

    if period == "anniversary":
        if anchor is None:
            return _calendar_month_window(now_date)
        return _anniversary_window(now_date, anchor.date())

    raise ValueError(f"unrecognized allowance_period: {period!r}")
